src/main/java/servlets/ReceiptAnalysis.java

```diff
@@ -189,14 +205,91 @@ private static AnalysisResults analyzeImage(ByteString imageBytes)
     return categories;
   }
 
-  /*
+  /**
    * Parse category strings into more natural categories
    * e.g. "/Food & Drink/Restaurants" becomes "Food & Drink" and "Restaurants"
    */
   private static Stream<String> parseCategory(ClassificationCategory category) {
     return Stream.of(category.getName().substring(1).split("/"));
   }
 
+  /**
+   * Checks the provided tokens for a date that can be parsed. If one is found, it is added to the
+   * builder as a timestamp.
+   */
+  private static void checkForParsableDate(
+      AnalysisResults.Builder analysisBuilder, Stream<String> tokens) {
+    Stream<String> dates = tokens.filter(ReceiptAnalysis::isDate);
+    // Assume that the first date on the receipt is the transaction date.
+    Optional<String> firstDate = dates.findFirst();
+
+    if (firstDate.isPresent()) {
+      addDateIfValid(analysisBuilder, firstDate.get());
+    }
```

Comment:
`Optional.ifPresent()` could be used here.

https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html#ifPresent-java.util.function.Consumer-